### ISSUE TITLE
changed class names

### ISSUE THREAD

**Issue by kostyfisik** [dealii/dealii#3218]
I am not sure that there is any possible solution for this issue, but it is rather confusing for a first time when watching video lectures and dealing with tutorial program it turns out that they are using different names for the same class.  E.g. (if I get it right, not sure) HyperShellBoundary was renamed to SphericalManifold, CompressedSparityPattern was renamed to DynamicSparsityPattern and so on. It is not obvious why it was renamed and it is completely confusing if you try to find such a name from the video lecture in some step-XX description using full-text search (it doesn`t match, you check the spelling, it still has no matches, so you start investigating the problem instead of learning deal.ii and FEM).

Probably it is possible to add a warning (at least for first six steps) after any video lecture mentioning that some names were changed (ideally linking to the file, where you can find the history of every class renaming).

**Comment by davydden** [comment]
The source of this confusion is that lectures have been recording over a long period of time. I personally watched them some 3-4 years ago, the first one probably was recorder much earlier. Obviously `deal.II` evolve and some functions are depricated, some are renamed, see for example https://www.dealii.org/developer/doxygen/deal.II/changes_after_8_4_2.html for the current list in `Incompatibilities`. This is natural a behaviour.

As for the solution, one could add notes/captions to video lectures (doable in Youtube and AFAIK Vimeo). So when @bangerth names any class that was renamed/depricated in the current release, there could be a caption `CompressedSparityPattern is renamed to DynamicSparsityPattern` or `CompressedSparityPattern is depricated, use DynamicSparsityPattern`.

**Comment by oneliefleft** [comment]
> As for the solution, one could add notes/captions to video lectures
> (doable in Youtube and AFAIK Vimeo). So when @bangerth names any
> class that was renamed/depricated in the current release, there could
> be a caption `CompressedSparityPattern is renamed to
> DynamicSparsityPattern` or `CompressedSparityPattern is depricated,
> use DynamicSparsityPattern`.

Smart idea.   :-)

**Comment by kostyfisik** [comment]
I think that maintaining steps` text should be easier compared to video annotation. And I am not sure that they will survive after downloading (actually I have  downloaded all the lectures to watch them during the flight) 

Incompatibilities list seems to be a bit scaring (at list in the context of tutorial program). I believe that the list of classes with changed names, listed in video lectures, is much shorter.

**Comment by davydden** [comment]
>  think that maintaining steps` text should be easier compared to video annotation. 

oh, you mean some text in tutorials is outdated? Make sure you use the consistent version of documentation and deal.II itself, i.e. `8.4.0` or `master`. If you found some mistakes in tutorials, please file a separate bug report with details so that we can fix it. Or you can also prepare a PR to fix the documentation yourself :smile:

> And I am not sure that they will survive after downloading

depends on how it is done. If it is hard-coded, it will survive. It is only an overlay in Youtube, then it won't.

**Comment by kostyfisik** [comment]
See the PR above.  Putting changes to the video files is a hard (and probably wrong) way. 

A complete solution can be, e.g. adding to  @dealiiVideoLecture{5,6} and @dealiiVideoLectureSeeAlso{2,4} compiled text of " and info" to the end, where info is a link to a special page, with renames listed for each step.

**Comment by davydden** [comment]
> See the PR above. Putting changes to the video files is a hard (and probably wrong) way.

if we don't hardcode them but only make annotations, it's quite [straightforward on Youtube](https://support.google.com/youtube/answer/92710?hl=en)

**Comment by kostyfisik** [comment]
Youtube edits leads to having the same docs in several places (which is not a good practice). At the same time, the "info" file i mentioned above can be also useful to update some legacy code that was linked to some old version of deal.ii. So the step-XX relevant data should be placed in the top of this file, all other renames - afterwards...

**Comment by bangerth** [comment]
As @davydden already pointed out, these videos have been continuously produced for several years now. The issue of incompatibilities has been on my mind for the entire time, but I have not come up with a good solution -- we *do* want to occasionally break backward compatibility if we encounter things that we now realize can be done better in a different way.

Each lecture has its own page, for example http://www.math.colostate.edu/~bangerth/videos.676.1.html . Rather than put a blanket statement somewhere that is generic and non-specific, I would much rather like to point out specific incompatibilities on these pages. I would definitely like to get feedback which class/function you found in which lecture and that no longer exists with that name. I would gladly list that in the text that accompanies the lecture.

**Comment by bangerth** [comment]
Ah, the thread is back :-) Welcome back, @kostyfisik !

**Comment by bangerth** [comment]
Yes, I know that you can do that. I just don't know anything I can do about it. I'm not the person you uploaded the videos to youtube, so don't "own" them, and can't change them. 

The only option I see is to point out incompatibilities in the pages I mentioned, or on the steps pages like you do in #3219.

**Comment by kostyfisik** [comment]
It was an unexpected experience of being excluded from GitHub for several hours :) It was quite useful to check that I have a local copy of all my GitHub repos..

> The only option I see is to point out incompatibilities in the pages I mentioned, or on the steps pages like you do in #3219.

We are talking about the same thing! :) The first lecture was published on the March, 2013, so it is around commit d02396c1674da3a7c8cf4c67ac19b26805474482  Mar 9, 2013 , so using in dealii root folder  
$git diff --color-words d02396c1674da3a7c8cf4c67ac19b26805474482:deal.II/examples/step-1/step-1.cc examples/step-1/step-1.cc |grep -v //
 
step-1

<deal.II/grid/tria_boundary_lib.h> ===>    <deal.II/grid/manifold_lib.h>
HyperShellBoundary<2> boundary_description  ===> SphericalManifold<2> manifold_description
triangulation.set_boundary(0, boundary_description)  ===> triangulation.set_manifold (0, manifold_description);

step-2 has changes from step-1 and 
<deal.II/lac/compressed_sparsity_pattern.h> ===> <deal.II/lac/dynamic_sparsity_pattern.h>
CompressedSparsityPattern compressed_sparsity_pattern ===> DynamicSparsityPattern dynamic_sparsity_pattern
print_gnuplot ===> print_svg

step-3 some from the previous and
std::vector<unsigned int>numbers  ===>  std::vector<types::global_dof_index> local_dof_indices 
q_point ===> q_index
std::map<unsigned int,double>  ===> std::map<types::global_dof_index,double> boundary_values;

step-4 - no new renames?

step-5

template <int dim>
class Coefficient : public Function<dim>  ===> removed, use
 template <int dim>
double coefficient (const Point<dim> &p)

step-6 - no new changes?

**Comment by bangerth** [comment]
I don't think we came to a conclusion here. There are three places we could potentially change:

1. The youtube video. This does not work since I don't own the videos.
2. The page that describes each of these videos
3. The tutorial itself. This is slightly awkward because it would mean that the tutorial program pages don't just document the current state, but also a past state and/or an outdated version of the program and the library. It also requires people who watch the lecture to *also* look at the tutorial page, rather than just working with the program.

So that leaves option 2. It would require me to keep a list of things that are outdated with each video lecture, but that I could do. I could use help identifying things that are outdated, though.

**Comment by tjhei** [comment]
> The youtube video. This does not work since I don't own the videos.

This is unfortunate. I wonder if this can be changed? You could reupload them and ask them to take the original ones down? Or, at least get control if you record new lectures in the future?

Note that while you can not edit the description of the videos, you could at least post a comment to the video that links back to the page where you keep track of the changes.

> So that leaves option 2. It would require me to keep a list of things that are outdated with each video lecture, but that I could do.

Yes, that sounds good.

**Comment by kostyfisik** [comment]
I believe that option 2 can be simplified. Just add a note that current state of the tutorial program file can differ from what is presented in the video. Probably providing a command to see the changes compared to the current version can be enough (see git diff command example it in my previous comment)

**Comment by bangerth** [comment]
@tjhei -- the Texas A&M TV studio owns them. They wanted it that way because they needed to track usage for their internal statistics. That was a fair request. Be that as it may, I'd like people to view them through my pages for multiple reasons (because they are cross-linked with other stuff, have the slide PDFs, ...) so having the information there as opposed to youtube is ok with me. It would be nice if I could link from youtube to the individual pages I have, but that's not all that important.

I'll have to create what changes on occasion. I'll put it somewhere on my TODO list, but it's not going to be high up I'm afraid.
